### PR TITLE
JAPI-342 Can't write DATAXX fields with dfsclient

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added support for writing fixed length data fields
- Fixed HPCCRecord comparison of byte arrays
- Corrected record translation to maintain fixed length field types

Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana @drealeed please review

**Comment by drealeed** [comment]
I tried testing with my tardis test  but this branch is missing the fix I put in last Friday (checking for BigDecimal when writing Integer fields in BinaryRecordWriter.)

**Comment by drealeed** [comment]
Also should this be merged to master or to candidate-7.12.x?

**Comment by rpastrana** [comment]
@jpmcmu please re-target candidate-7.12.x.

**Review by drealeed — APPROVED** [review]
Tested with the tardis module and the issue's fixed.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java [hunk 1]
```diff
@@ -266,10 +266,39 @@ private void writeField(FieldDef fd, Object fieldValue) throws Exception
...
-                long dataSize = data.length;
-                writeUnsigned(dataSize);
-                writeByteArray(data);
+                if (fd.isFixed())
+                {
+                    byte[] data = fieldValue==null?new byte[0]:(byte[]) fieldValue;
+                    int bytesToWrite = data.length;
+                    if (bytesToWrite > fd.getDataLen())
+                    {
+                        bytesToWrite = (int) fd.getDataLen();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:275]: some data is being cut off? does it make sense to log?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java [hunk 2]
```diff
@@ -266,10 +266,39 @@ private void writeField(FieldDef fd, Object fieldValue) throws Exception
...
+                        }
+
+                        Arrays.fill(scratchBuffer, 0, fillLength, (byte) '\0');
+                        writeByteArray(scratchBuffer, 0, fillLength);
+                        numFillBytes -= fillLength;
+                    }
+                }
+                else
+                {
+                    byte[] data = fieldValue==null?new byte[0]:(byte[]) fieldValue;
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:297]: minor but this line is repeated, if it doesn't affect maintainability lets move out of the conditional

**Review by rpastrana — COMMENTED** [review]
@jpmcmu looks good, a couple of minor comments

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java (hunk 1 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:275]: Good call, I will add that in

**Comment by jpmcmu** [comment]
@rpastrana Added logging for truncation case

**Comment by jpmcmu** [comment]
@rpastrana added automatic output of messages in close

**Review by rpastrana — APPROVED** [review]
@jpmcmu looks good

**Pull request merged**